networkx/algorithms/centrality/knotty.py

```diff
@@ -0,0 +1,165 @@
+                        'directed or multiedge graphs.')
+    return _find_knotty_centre(G, compact)
+
+
+def _find_knotty_centre(G, compact):
+    # nodes = the sub-graph found
+    # kc = its knotty-centredness
+
+    N = G.number_of_nodes()
+
+    # binarise matrix - all non-zero weights become 1s
+    edge_matrix = np.array(nx.to_numpy_matrix(G))
+    CIJ = edge_matrix!=0
+    CIJ = CIJ.astype(int)
+
+
+    # Exhastive search phase
+    Exh = 5 # number of nodes for exhaustive search (2^Exh combinations)
+    Exh = min([Exh,N])
+    
+    BC_dict = nx.betweenness_centrality(G)
+    BC_list = list(BC_dict.values())
+    
+    BC = np.array(BC_list)/sum(BC_list) # normalise wrt total betweenness centrality
+
+    # Calculate indirect betweenness centrality
+    BC2 = BC + np.sum(CIJ*BC,axis=0) + np.sum(CIJ,axis=1).T*BC
+
+    # Ranking the nodes in descending order
+    array_asc = (BC2.ravel().argsort())
+    list_asc = array_asc.tolist()
+    list_asc.reverse() #list_asc is not descending
+    IxBC = list_asc
+    
+    nodes = []
+    improving = 1
+    
+    while improving:
+        L = len(nodes)
+        nodes_left = IxBC
+        nodes_left = [x for x in nodes_left if x not in set(nodes)]
+        choices = [nodes_left[i] for i in range(0, min([Exh,len(nodes_left)]))]
+        nodes, kc = _best_perm(nodes, choices, G, CIJ, compact, BC)
+        improving = len(nodes) > L
+
+    # Hill climbing phase
+    nodes_left = range(0,N)
+    nodes_left = [x for x in nodes_left if x not in set(nodes)]
+    improving = 1
+    
+    while improving and nodes_left:
+        best_kc = 0
+        for node in nodes_left:
+            nodes2 = np.hstack((nodes, node))
+            kc2 = _compute_knotty_centrality(G, CIJ, nodes2, compact, BC)
+            if kc2 > best_kc:
+                best_kc = kc2
+                best_node = node
+                
+        if best_kc > kc:
+            kc = best_kc
+            nodes = np.hstack((nodes, best_node))
+            nodes_left.remove(best_node)
+        else:
+            improving = 0
+    return nodes, kc
+
+def _best_perm(given, choices, G, CIJ, compact, BC):
+    # Carries out exhaustive search to find a permutation of nodes in
+    # "choices" that when added to the nodes in "given" yields the highest
+    # value of knotty-centrality
+
+    if choices:
+        choices2 = [choices[i] for i in range(0, len(choices))]
+        new = choices[0]
+        nodes1, kc1 = _best_perm(np.hstack((given, new)), choices2, G, CIJ, compact, BC)
+        nodes2, kc2 = _best_perm(given, choices2, G, CIJ, compact, BC)
+        if kc1 > kc2:
+            nodes = nodes1
+            kc = kc1
+        else:
+            nodes = nodes2
+            kc = kc2
+    else:
+        nodes = given
+        kc = _compute_knotty_centrality(G, CIJ, nodes, compact, BC)
+    return nodes, kc
+
+def _compute_knotty_centrality(G, CIJ, nodes, compact, BC):
+    # Returns knotty-centrality of the subgraph of CIJ comprising only
```

Comment:
Passing `CIJ` around seems largely wasteful here -- it looks like it's only being used to get the size `n = len(G)`, after a quite expensive "binarise" step which redundant because that happens at the top level.

In retrospect, doing the powerset in a recursive (or pseudo-recursive to avoid blowing the stack limit) manner would be good, because `BCtot` and `density` could be done in constant / `n` time respectively instead of `n` and `n^2` time per call.